Improved form rendering

madsmtm opened the pull request:
Updated form rendering to properly use bootstrap v.4, it now also works with nested forms and custom form elements.
Basically, I just went to [bootstrap/components/forms](https://getbootstrap.com/docs/4.1/components/forms/), and implemented the rendering based on the examples there.
I've made the interface a lot more flexible, but and if if you think its needed yet, I can make it backwards compatible

If you approve, I'll attempt to make some documentation

madsmtm commented:
Since you can't easily override macros, I've wrapped them all in a block that you can then extend, the alternative would be [this](https://stackoverflow.com/a/26630528). There's also a block that you can extend if you need to provide rendering of custom form elements

madsmtm commented:
I'm still not really satisfied with how widgets, especially `TableWidget`s, are rendered inside `ListWidget`s, I'd like the label & help text to be in a top row if possible

greyli commented:
Thanks very much! I will review your code as soon as I have free time.

madsmtm commented on flask_bootstrap/templates/bootstrap/form.html:
```diff
@@ -1,231 +1,227 @@
+
+{%- macro hidden_errors(field) -%}
+{%- block hidden_errors scoped -%}
+    {# Doesn't actually work, has no element which makes it "display: block" #}
```
Also regarding this, do you know how the preffered way of displaying errors like this is in bootstrap? We can't use the normal `<span class="invalid-feedback">`, since that's specific to form elements. Something like `<div class="flash alert alert-danger">` perhaps?

greyli replied:
Use Alert will be good. Why adding a `flash` class?

madsmtm replied:
I'm not really sure why I thought that, I probably just looked at some legacy code ;)

greyli commented on setup.py:
```diff
@@ -32,8 +32,11 @@
     test_suite='test_bootstrap_flask',
     install_requires=[
-        'Flask'
+        'Flask>=1'
```
Why need Flask **>= 1**?

madsmtm replied:
Because `Flask` versions below `v1.0.0` require `Jinja2>=2.4`, while we need at least `Jinja2>=2.8` (Or around there). Maybe it would be better to add an explicit requirement?

greyli replied:
All right...

greyli commented:
LGTM! Just go ahead to update the docs.

greyli commented:
> I've made the interface a lot more flexible, but and if you think its needed yet, I can make it backward compatible.

Does this PR break something?

madsmtm commented:
> Does this PR break something?
I've changed the interface of the `render_field`, `render_form` macros and removed the `form_errors` macro. I've also removed the `bootstrap_is_hidden_field` jinja function, and replaced it with the `is_hidden` test

greyli commented:
IMO, it's a good idea to make it backward compatible.

madsmtm commented:
I've tried documenting a little bit, but I'm unsure about the format. Should I:
- Make a seperate file, where I put information about all the methods, and how to customize the form rendering
- Only document the single `render_form` method

Or something in between those two?

greyli commented on docs/macros.rst:
```diff
@@ -120,51 +84,20 @@ Example
 
-.. py:function:: quick_form(form,\
-                    action="",\
-                    method="post",\
```
Why these arguments were removed?

madsmtm replied:
I made them be part of `kwargs` instead

greyli commented:
Can we just keep the old keyword arguments (i.e. `action`, `method`, `extra_classes` etc)? Both in the code and the docs.

madsmtm commented:
If you feel like we should do that, then yes, easily. Which ones do you want to keep precisely, `action`, `method`, `role`, `enctype`, `id`, `novalidate` and `extra_classes`?

greyli commented:
@madsmtm We need to let user transfer easily from Flask—Bootstrap to this project, so I want to keep all of them.

madsmtm commented:
Well, but I don't see a reason to update the code, since after I made it backwards compatible, they can, they don't need to do anything. But if you want it added to the docs, I'll be happy to do so, I'm out of time for today though

greyli commented on examples/templates/form.html:
```diff
@@ -2,15 +2,13 @@
-    {{ render_field(form.password) }}
-    {{ render_field(form.remember) }}
-    {{ render_field(form.submit) }}
-    </form>
```
Why delete this example? People sometimes need to render the field one by one.

madsmtm replied:
Yes, I'll move it into the seperate documentation file

greyli commented:
OK, you are right, just update the docs will be enough.

PanderMusubi commented:
Looking forward too seeing this in the next version. :+1:

PanderMusubi commented:
@madsmtm would you like (with our help) finish this PR?
